### PR TITLE
event_base: remove event_base ptr from FileEventImpl

### PR DESCRIPTION
**Pull request by lambdai**
Inspired by my experiment fighting with event.
envoy never migrate from FileEvent to another event_base. Notes the ownership of fd could be moved.
Since `base_` always holds the value of `event_get_base(&raw_event_)`, remove the `base_` field and save 8 bytes on major 64bit platforms.

### CODE REVIEW

File: source/common/event/file_event_impl.cc [hunk 1]
```diff
@@ -39,7 +40,14 @@ void FileEventImpl::activate(uint32_t events) {
...
 }
 
+void FileEventImpl::assignEventBase(event_base* base) {
+  // we cannot use below
+  // ASSERT(-1 !=event_base_set(base, &raw_event_));
+  raw_event_.ev_base = base;
+}
+
 void FileEventImpl::assignEvents(uint32_t events) {
+  ASSERT(base_ == event_get_base(&raw_event_));
```
Comment (lambdai) [inline source/common/event/file_event_impl.cc]: @mattklein123 I want to demonstrate that `FileEventImpl::base_` is not required. 
The reason is that `base_` could be extract from `raw_event_`.
So we can remove the member `base_` from `FileEventImpl` and get 8 bytes back.

Comment (mattklein123) [reply source/common/event/file_event_impl.cc]: Yes, OK, this makes sense, though I would implement it slightly differently vs. directly accessing the ev_base member. Basically just modify assignEvents() to take a base as a parameter. The first time it's called, pass the base from the constructor, then in `setEnabled()` pass the base from the event itself.

Please update the PR title, description, etc. to be more clear. Thanks.

/wait

Comment (lambdai) [reply source/common/event/file_event_impl.cc]: SG. Will follow your suggestion and update. 
Thanks!

File: source/common/event/file_event_impl.cc [hunk 2]
```diff
@@ -68,8 +69,10 @@ void FileEventImpl::assignEvents(uint32_t events) {
 }
 
 void FileEventImpl::setEnabled(uint32_t events) {
+  auto* base = event_get_base(&raw_event_);
+  ASSERT(base != nullptr);
```
Comment (mattklein123) [inline source/common/event/file_event_impl.cc]: You ASSERT this in assignEvents() so I think you can remove this.

Comment (lambdai) [reply source/common/event/file_event_impl.cc]: Sharp eye!

Comment (lambdai) [reply source/common/event/file_event_impl.cc]: sharp eye!

Comment (lambdai) [reply source/common/event/file_event_impl.cc]: sharp eye!

**Pull request merged**